Enable multiple concurrent consumers for a given queue

oscarfh opened the pull request:
This fixes #248.
I wish I could add this config as an application.yml configuration, but the queue config seems to be done in the `@Queue` annotation, so I followed the pattern.

Now one can configure the `numberOfConsumers` in the `@Queue` annotation so we have multiple consumers. I see it being used with `numberOfConsumers`== `consumers.nThreads`, so there is a 1:1 relation between consumers and threads.

The prefetch count will still be individual to each consumer.

jameskleeh commented on rabbitmq/src/main/java/io/micronaut/rabbitmq/connect/ChannelPool.java:
```diff
@@ -40,6 +41,16 @@
+     * @return The list of channels
+     * @throws IOException If there were errors creating any of the channels
+     */
+    List<Channel> getMultipleChannels(int numberOfChannels) throws IOException;
```
I think we can do without this API. You can just call getChannel n times.

jameskleeh commented on rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java:
```diff
@@ -128,180 +128,182 @@ public void process(BeanDefinition<?> beanDefinition, ExecutableMethod<?, ?> met
-                if (prefetch != null) {
-                    channel.basicQos(prefetch);
+            for (Channel channel : channelList) {
+                Integer prefetch = queueAnn.get("prefetch", Integer.class).orElse(null);
```
This can be retrieved outside the loop

jameskleeh commented on rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java:
```diff
@@ -128,180 +128,182 @@ public void process(BeanDefinition<?> beanDefinition, ExecutableMethod<?, ?> met
+                    }
 
-                channel.basicConsume(queue, false, clientTag, false, exclusive, arguments, new DefaultConsumer() {
+                    channel.basicConsume(queue, false, clientTag, false, exclusive, arguments, new DefaultConsumer() {
```
Most of the above logic can be done outside the loop

jameskleeh reviewed (changes requested):
This introduces some inefficiencies because work is done inside the loop when it could be done outside

jameskleeh commented:
The consumer instance (new DefaultConsumer) should also be created outside of the loop

oscarfh commented:
@jameskleeh Thanks for the feedback I incorporated them and the code is much better now. I extracted some methods to makes things easier to understand, I hope this is also OK.
I am not sure I get your comment about moving the creation of the `DefaultConsumer` to outside of the loop. The `DefaultConsumer` manipulates the channels, specially ack'ing messages and returning the channel to the pool. We could have one single `DefaultConsumer` that handles all channels, but in this case we would need to match the message to the channel used to receive it, so we can manipulate it correctly.
To me, it seems safer to have one `DefaultConsumer` per channel (current approach), so it is less error prone, what do you think?

jameskleeh commented:
OK that should be fine. Will review tomorrow

On Sun, Jul 11, 2021 at 8:26 PM oscarfh ***@***.***> wrote:

> @jameskleeh <https://github.com/jameskleeh> Thanks for the feedback I
> incorporated them and the code is much better now. I extracted some methods
> to makes things easier to understand, I hope this is also OK.
> I am not sure I get your comment about moving the creation of the
> DefaultConsumer to outside of the loop. The DefaultConsumer manipulates
> the channels, specially ack'ing messages and returning the channel to the
> pool. We could have one single DefaultConsumer that handles all channels,
> but in this case we would need to match the message to the channel used to
> receive it, so we can manipulate it correctly.
> To me, it seems safer to have one DefaultConsumer per channel (current
> approach), so it is less error prone, what do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/micronaut-projects/micronaut-rabbitmq/pull/249#issuecomment-877888193>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMCVLOMTQ2BEBCI6XXHMLLTXIZCXANCNFSM5AC7NMYQ>
> .
>

mscheibler commented on rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java:
```diff
@@ -128,180 +130,203 @@ public void process(BeanDefinition<?> beanDefinition, ExecutableMethod<?, ?> met
-                    .map(type -> Qualifiers.byAnnotation(beanDefinition, type))
-                    .orElse(null);
+            Map<String, Object> arguments = retrieveArguments(method);
+            Object bean = getExecutableMethodBean(beanDefinition, method);
```
As i did a review of the code i had a concern about thread safety.
By using multiple channels and threads to consume messages from a queue, the `@RabbitListener` bean implementation must be thread-safe.
With the current implementation, with only one thread per consumer bean this is not an issue.

Maybe it is sufficient to mention it in the documentaion of `@RabbitListener`.
Another possibility would be to have a separate `@RabbitListener` bean instance per consumer/thread so that the bean itself does not need to care about it. But then this issue just traverse to other beans used by the rabbit listener.

Do i see ghosts or is this something to consider?

oscarfh replied:
This is a very good point. AFAIK, the `@RabbitListener` is a `Singleton`, so this is already a requirement. I do not know whether we can influence the creation of beans from this class, or whether we can remove the characteristic of it being a Singleton.

jameskleeh replied:
Given the feature is opt in, I don't see that as an issue. If you enable concurrent execution, then one should be aware any manipulation of data on the bean is thread safe. It's worth a note in the documentation though.

jameskleeh commented:
Did some cleanup and added docs here https://github.com/micronaut-projects/micronaut-rabbitmq/pull/250
